formatter/src/main/java/org/togetherjava/formatter/tokenizer/Lexer.java

```diff
@@ -0,0 +1,92 @@
+package org.togetherjava.formatter.tokenizer;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Tokenizer that can turn a list of strings (or a string) into a list of tokens
+ *
+ * @author illuminator3
+ */
+public class Lexer {
+    private static final Pattern commentPatcherRegex =
+            Pattern.compile("(/\\*[^*]*\\*+(?:[^/*][^*]*\\*+)*/)");
+
+    /**
+     * Tokenizes the given input by tokenizing each line individually (splitting by \n)
+     *
+     * @param input input to tokenize
+     * @return resulting tokens
+     * @author illuminator3
+     */
+    public List<Token> tokenize(String input) {
+        return tokenize(Arrays.asList(patchComments(input).split("\n")));
+    }
+
+    /**
+     * Tokenizes the given input line by line
+     *
+     * @param lines input to tokenize
+     * @return resulting tokens
+     * @author illuminator3
+     */
+    public List<Token> tokenize(List<String> lines) {
+        return lines.stream().map(this::tokenizeLine).flatMap(List::stream).toList();
+    }
+
+    /**
+     * Tokenizes a single line using regex
+     *
+     * @param line input to tokenize
+     * @return resulting tokens
+     * @author illuminator3
+     */
+    private List<Token> tokenizeLine(String line) {
+        List<Token> tokens = new ArrayList<>();
+        int index = 0;
+        String content;
+
+        while (!(content = line.substring(index)).isEmpty()) {
+            tokenCheck: {
```

Comment:
Why do you have a random block here with a name? Why not just use `break`? It currently is harder to read than necessary IMHO.

You would need a nested break here, which indicates that the inner loop should probably be a method somewhere and not inlined here. Maybe sth like:
```java
while (!(content = line.substring(index)).isEmpty()) {
    // can throw or return an optional and you throw in here
    Token matchingToken = findTokenInString(content);
    index += matchingToken.length();
    tokens.add(matchingToken);
}
```